### PR TITLE
ref(spans): Use an SQL parser to normalize span descriptions

### PR DESCRIPTION
**Pull request by jjbayer**
This will allow us to implement more complex normalizations, such as collapsing `CASE WHEN` expressions.

#skip-changelog

### CODE REVIEW

File: relay-event-normalization/src/normalize/span/description/sql/mod.rs [hunk 1]
```diff
@@ -208,112 +231,126 @@ mod tests {
...
+        r#"SELECT * from table1 WHERE sku = "foo""#,
+        "SELECT * FROM table1 WHERE sku = foo"
     );
 
     scrub_sql_test!(
         num_where,
-        "SELECT * FROM table WHERE id = 1",
-        "SELECT * FROM table WHERE id = %s"
+        "SELECT * FROM table1 WHERE id = 1",
+        "SELECT * FROM table1 WHERE id = %s"
```
Comment (jjbayer) [inline relay-event-normalization/src/normalize/span/description/sql/mod.rs:335]: I changed most occurences of unquoted `table` because it is not valid SQL.

File: relay-event-normalization/src/normalize/span/description/mod.rs [hunk 2]
```diff
@@ -30,7 +31,13 @@ pub(crate) fn scrub_span_description(span: &mut Span, rules: &Vec<SpanDescriptio
...
             ("http", _) => scrub_http(description),
             ("cache", _) | ("db", "redis") => scrub_redis_keys(description),
-            ("db", _) => sql::scrub_queries(description),
+            ("db", _) => sql::scrub_queries(
+                span.data
+                    .value()
+                    .and_then(|d| d.get("system"))
+                    .and_then(|v| v.as_str()),
+                description,
+            ),
```
Comment (jjbayer) [inline relay-event-normalization/src/normalize/span/description/mod.rs:40]: The "db system" from the span is now used to inform the query scrubber about the SQL dialect.

File: relay-event-normalization/src/normalize/span/description/sql/mod.rs [hunk 3]
```diff
@@ -24,7 +29,7 @@ static NORMALIZER_REGEX: Lazy<Regex> = Lazy::new(|| {
         # Capture single-quoted strings, including the remaining substring if `\'` is found.
         ((?-x)(?P<single_quoted_strs>'(?:\\'|[^'])*(?:'|$)(::\w+(\[\]?)?)?)) |
         # Capture placeholders.
-        (   (?P<placeholder> (?:\?+|\$\d+|%s|:\w+) (::\w+(\[\]?)?)? )   ) |
+        (   (?P<placeholder> (?:\?+|\$\d+|%(?:\(\w+\))?s|:\w+) (::\w+(\[\]?)?)? )   ) |
```
Comment (jjbayer) [inline relay-event-normalization/src/normalize/span/description/sql/mod.rs:32]: This is a drive-by to support placeholders like `%(name)s`.

File: relay-event-normalization/src/normalize/span/description/sql/mod.rs [hunk 4]
```diff
@@ -87,7 +92,21 @@ static COLLAPSE_COLUMNS: Lazy<Regex> = Lazy::new(|| {
 static ALREADY_NORMALIZED_REGEX: Lazy<Regex> = Lazy::new(|| Regex::new(r"/\?|\$1|%s").unwrap());
 
 /// Normalizes the given SQL-query-like string.
-pub(crate) fn scrub_queries(string: &str) -> Option<String> {
+pub fn scrub_queries(db_system: Option<&str>, string: &str) -> Option<String> {
+    let t = Instant::now();
+    let (res, mode) = scrub_queries_inner(db_system, string);
+    relay_statsd::metric!(
+        timer(Timers::SpanDescriptionNormalizeSQL) = t.elapsed(),
+        mode = mode
```
Comment (iker-barriocanal) [inline relay-event-normalization/src/normalize/span/description/sql/mod.rs]: Should we make `mode` an enum to avoid future high cardinality problems?

File: relay-event-normalization/src/normalize/span/description/sql/parser.rs [hunk 5]
```diff
@@ -0,0 +1,274 @@
...
+use std::ops::ControlFlow;
+
+use itertools::Itertools;
+use sqlparser::ast::{
+    Expr, Ident, Query, SelectItem, SetExpr, Statement, TableFactor, UnaryOperator, Value,
+    VisitMut, VisitorMut,
+};
+use sqlparser::dialect::{Dialect, GenericDialect};
+
+/// Derive the SQL dialect from `span.system` and try to parse the query into an AST.
```
Comment (iker-barriocanal) [inline relay-event-normalization/src/normalize/span/description/sql/parser.rs]: The value of `span.system` is now provided as an argument to the function.

File: relay-event-normalization/src/normalize/span/description/sql/parser.rs [hunk 6]
```diff
@@ -0,0 +1,274 @@
...
+
+    /// Placeholder for a list of items that has been scrubbed.
+    fn ellipsis() -> Ident {
+        Ident::new("..")
+    }
+
+    /// Check if a selected item can be erased into `..`. Currently we only collapse simple
+    /// columns (`col1` or `col1 AS foo`).
+    fn is_collapsible(item: &SelectItem) -> bool {
+        match item {
```
Comment (iker-barriocanal) [inline relay-event-normalization/src/normalize/span/description/sql/parser.rs:57]: Should we also collapse `SelectItem::QualifiedAlias`?

File: relay-event-normalization/src/normalize/span/description/sql/parser.rs [hunk 7]
```diff
@@ -0,0 +1,274 @@
...
+            Expr::CompoundIdentifier(parts) => {
+                Self::simplify_compound_identifier(parts);
+            }
+            // `"col"` is replaced by `col`.
+            Expr::Identifier(ident) => {
+                ident.quote_style = None;
+            }
+            // Recurse into subqueries.
+            Expr::Subquery(query) => Self::transform_query(query),
+            // Remove negative sign, e.g. `-100` becomes `%s`.
```
Comment (iker-barriocanal) [inline relay-event-normalization/src/normalize/span/description/sql/parser.rs:160]: IIUC this doesn't remove the negative sign, but just resets the expression (i.e. `-100` becomes `-%s`)? See https://github.com/sqlparser-rs/sqlparser-rs/blob/4c3a4ad5a87ef8b4dd24a6e8540243e14fbfa148/src/ast/mod.rs#L729

File: relay-server/src/metrics_extraction/snapshots/relay_server__metrics_extraction__event__tests__extract_span_metrics.snap [hunk 8]
```diff
@@ -604,7 +604,9 @@ expression: metrics
             "http.status_code": "500",
             "span.action": "INSERT",
             "span.category": "db",
+            "span.description": "INSERT INTO from_date (..) VALUES (%s)",
```
Comment (iker-barriocanal) [inline relay-server/src/metrics_extraction/snapshots/relay_server__metrics_extraction__event__tests__extract_span_metrics.snap:607]: Are we using `(..)` on replacements, or just `..`? Can this result in unexpected aggregation outcomes?

File: relay-event-normalization/src/normalize/span/description/sql/parser.rs [hunk 9]
```diff
@@ -0,0 +1,274 @@
...
+                output.append(collapse);
+            }
+            _ => {
+                output.push(SelectItem::UnnamedExpr(Expr::Identifier(Self::ellipsis())));
+            }
+        }
+    }
+
+    /// Normalizes `SELECT ...` queries.
+    fn transform_query(query: &mut Query) {
```
Comment (iker-barriocanal) [inline relay-event-normalization/src/normalize/span/description/sql/parser.rs:82]: nit: I'd split this function into different ones for each keyword (e.g. select, from).

**Review by iker-barriocanal — APPROVED** [review]
Good stuff, so clean! ~It'd be great if we can add tests for these complex cases, like the `CASE WHEN` in the description.~ just saw the other PR

File: relay-event-normalization/src/normalize/span/description/sql/parser.rs (hunk 6 above)
Comment (jjbayer) [reply relay-event-normalization/src/normalize/span/description/sql/parser.rs:57]: Is that a thing? I only see an `QualifiedWildcard` variant, but that's different: https://github.com/sqlparser-rs/sqlparser-rs/blob/4c3a4ad5a87ef8b4dd24a6e8540243e14fbfa148/src/ast/query.rs#L383-L392.

File: relay-event-normalization/src/normalize/span/description/sql/parser.rs (hunk 5 above)
Comment (jjbayer) [reply relay-event-normalization/src/normalize/span/description/sql/parser.rs]: 👍 Updated the comment.

File: relay-event-normalization/src/normalize/span/description/sql/parser.rs (hunk 9 above)
Comment (jjbayer) [reply relay-event-normalization/src/normalize/span/description/sql/parser.rs:82]: Done.

File: relay-server/src/metrics_extraction/snapshots/relay_server__metrics_extraction__event__tests__extract_span_metrics.snap (hunk 8 above)
Comment (jjbayer) [reply relay-server/src/metrics_extraction/snapshots/relay_server__metrics_extraction__event__tests__extract_span_metrics.snap:607]: In this case `(` and `)` are an essential part of [SQL syntax](https://www.postgresql.org/docs/current/sql-insert.html) so I would leave them.

File: relay-event-normalization/src/normalize/span/description/sql/parser.rs (hunk 7 above)
Comment (jjbayer) [reply relay-event-normalization/src/normalize/span/description/sql/parser.rs:160]: This works because it replaces the _outer_ expression. See https://github.com/getsentry/relay/blob/1ec03d82da20fbf30df69c88a2b3579cb35adf39/relay-event-normalization/src/normalize/span/description/sql/mod.rs#L344-L348.

File: relay-event-normalization/src/normalize/span/description/sql/mod.rs (hunk 4 above)
Comment (jjbayer) [reply relay-event-normalization/src/normalize/span/description/sql/mod.rs]: Done!

File: relay-event-normalization/src/normalize/span/description/sql/parser.rs (hunk 6 above)
Comment (iker-barriocanal) [reply relay-event-normalization/src/normalize/span/description/sql/parser.rs:57]: Hmm, I must have followed the wrong reference.

**Review by iker-barriocanal — APPROVED** [review]
🚀
